File: web/src/stores/setting.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025-2026 lin-snow

import { defineStore } from 'pinia'
import { ref } from 'vue'
import {
  fetchGetSettings,
  fetchGetS3Settings,
  fetchGetOAuth2Settings,
  fetchGetAllWebhooks,
  fetchListAccessTokens,
  fetchGetSnapshotScheduleSetting,
  fetchStartExport,
  fetchGetExportStatus,
  fetchGetAgentSettings,
  fetchGetAgentInfo,
  fetchHelloEch0,
} from '@/service/api'
import type { ExportStatusPayload } from '@/service/api'
import { S3Provider, OAuth2Provider, AgentProtocol } from '@/enums/enums'
import { useUserStore } from './user'

const SNAPSHOT_STATUS_POLL_INTERVAL_MS = 3000
type SnapshotUIStatus = 'idle' | 'pending' | 'running' | 'success' | 'failed' | 'cancelled'

export const useSettingStore = defineStore('settingStore', () => {
  const userStore = useUserStore()

  /**
   * State
   */
  const SystemSetting = ref<App.Api.Setting.SystemSetting>({
    site_title: import.meta.env.VITE_APP_TITLE,
    server_logo: '/Ech0.svg',
    server_logo_file_id: '',
    server_name: import.meta.env.VITE_APP_NAME,
    server_url: '',
    allow_register: true,
    default_locale: 'zh-CN',
    ICP_number: '',
    footer_content: '',
    footer_link: '',
    meting_api: '',
    custom_css: '',
    custom_js: '',
  })
  const S3Setting = ref<App.Api.Setting.S3Setting>({
    enable: false,
    provider: S3Provider.AWS,
    endpoint: '',
    access_key: '',
    secret_key: '',
    bucket_name: '',
    region: '',
    use_ssl: false,
    cdn_url: '',
    path_prefix: '',
    public_read: true,
  })
  const OAuth2Setting = ref<App.Api.Setting.OAuth2Setting>({
    enable: false,
    provider: OAuth2Provider.GITHUB,
    client_id: '',
    client_secret: '',
    redirect_uri: '',
    scopes: [],
    auth_url: '',
    token_url: '',
    user_info_url: '',

    is_oidc: false,
    issuer: '',
    jwks_url: '',
    auth_redirect_allowed_return_urls: [],
    cors_allowed_origins: [],
  })
  const Webhooks = ref<App.Api.Setting.Webhook[]>([])
  const webhooksLoading = ref<boolean>(false)
  const webhooksError = ref<string>('')
  const AccessTokens = ref<App.Api.Setting.AccessToken[]>([])
  const SnapshotSchedule = ref<App.Api.Setting.SnapshotSchedule>({
    enable: false,
    cron_expression: '0 2 * * 0',
  })
  const AgentSetting = ref<App.Api.Setting.AgentSetting>({
    enable: false,
    protocol: AgentProtocol.OPENAI,
    model: '',
    api_key: '',
    prompt: '',
    base_url: '',
    multimodal: false,
    context_window: 0,
  })
  const hello = ref<App.Api.Ech0.HelloEch0>()
  const loading = ref<boolean>(true)
  // 快照 = 导出作业（job.Manager，按类型单行、服务端持久化）。前端不再持有 taskId / localStorage，
  // 状态直接来自 GET /migration/export/status（与导入状态机一致）。
  const snapshotStatus = ref<SnapshotUIStatus>('idle')
  const snapshotError = ref<string>('')
  const snapshotPolling = ref<boolean>(false)
  const snapshotPollTimer = ref<ReturnType<typeof setTimeout> | null>(null)
  const snapshotPollInFlight = ref<boolean>(false)

  /**
   * Actions
   */
  const getSystemSetting = async () => {
    await fetchGetSettings().then((res) => {
      if (res.code === 1) {
        SystemSetting.value = res.data
        loading.value = false
      }
    })
  }

  const getS3Setting = async () => {
    fetchGetS3Settings().then((res) => {
      if (res.code === 1) {
        S3Setting.value = res.data
      }
    })
  }

  const getOAuth2Setting = async () => {
    const res = await fetchGetOAuth2Settings()
    if (res.code === 1) {
      OAuth2Setting.value = res.data
    }
    return OAuth2Setting.value
  }

  const getAllWebhooks = async () => {
    webhooksLoading.value = true
    webhooksError.value = ''
    try {
      const res = await fetchGetAllWebhooks()
      if (res.code === 1) {
        Webhooks.value = res.data || []
        return
      }
      webhooksError.value = res.msg || 'Failed to load webhooks'
    } catch (error) {
      webhooksError.value = error instanceof Error ? error.message : 'Failed to load webhooks'
    } finally {
      webhooksLoading.value = false
    }
  }

  const getAllAccessTokens = async () => {
    const res = await fetchListAccessTokens()
    if (res.code === 1) {
      if (res.data) {
        AccessTokens.value = res.data
      } else {
        AccessTokens.value = []
      }
    }
  }

  const getSnapshotSchedule = async () => {
    const res = await fetchGetSnapshotScheduleSetting()
    if (res.code === 1) {
      SnapshotSchedule.value = res.data
    }
  }

  const stopSnapshotPolling = () => {
    snapshotPolling.value = false
    if (snapshotPollTimer.value) {
      clearTimeout(snapshotPollTimer.value)
      snapshotPollTimer.value = null
    }
  }

  const applyExportState = (data: ExportStatusPayload) => {
    snapshotStatus.value = data.status
    snapshotError.value = data.error_message || ''
  }

  const isExportTerminal = (status: SnapshotUIStatus) =>
    status === 'idle' || status === 'success' || status === 'failed' || status === 'cancelled'

  const scheduleSnapshotPoll = () => {
    if (!snapshotPolling.value) return
    if (snapshotPollTimer.value) clearTimeout(snapshotPollTimer.value)
    snapshotPollTimer.value = setTimeout(() => {
      void pollSnapshotStatus()
    }, SNAPSHOT_STATUS_POLL_INTERVAL_MS)
  }

  const pollSnapshotStatus = async () => {
    if (snapshotPollInFlight.value) return
    snapshotPollInFlight.value = true
    try {
      const res = await fetchGetExportStatus()
      if (res.code === 1) {
        applyExportState(res.data)
        if (isExportTerminal(res.data.status)) {
          stopSnapshotPolling()
          return
        }
      }
    } catch (error) {
      snapshotError.value = error instanceof Error ? error.message : 'Failed to poll export status'
    } finally {
      snapshotPollInFlight.value = false
    }
    scheduleSnapshotPoll()
  }

  const startSnapshotTask = async () => {
    if (snapshotStatus.value === 'pending' || snapshotStatus.value === 'running') return null
    const res = await fetchStartExport()
    if (res.code === 1 && res.data) {
      applyExportState(res.data)
      snapshotPolling.value = true
      scheduleSnapshotPoll()
    }
    return res
  }

  // 进入页面时从服务端恢复：仅当存在进行中的导出作业才接管轮询；终态不回灌（避免每次挂载误弹成功提示）。
  const restoreSnapshotTask = async () => {
    const res = await fetchGetExportStatus()
    if (res.code === 1 && (res.data.status === 'pending' || res.data.status === 'running')) {
      applyExportState(res.data)
      snapshotPolling.value = true
      scheduleSnapshotPoll()
    }
  }

  const getHelloEch0 = async () => {
    const res = await fetchHelloEch0()
    if (res.code === 1) {
      hello.value = res.data
    }
  }

  const getAgentSetting = async () => {
    const res = await fetchGetAgentSettings()
    if (res.code === 1) {
      AgentSetting.value = res.data
    }
  }

  const getAgentInfo = async () => {
    const res = await fetchGetAgentInfo()
    if (res.code === 1) {
      AgentSetting.value.enable = res.data.enable
      AgentSetting.value.model = res.data.model
      AgentSetting.value.protocol = res.data.protocol
    }
  }

  const init = async () => {
    await getSystemSetting()
    if (userStore.isLogin && userStore.user?.is_admin) {
      getS3Setting()
    }
    getAgentInfo()
    getHelloEch0()
  }

  return {
    SystemSetting,
    S3Setting,
    OAuth2Setting,
    Webhooks,
    webhooksLoading,
    webhooksError,
    AccessTokens,
    SnapshotSchedule,
    AgentSetting,
    hello,
    loading,

    getAllAccessTokens,
    getSystemSetting,
    getS3Setting,
    getOAuth2Setting,
    getAllWebhooks,
    getHelloEch0,
    getSnapshotSchedule,
    startSnapshotTask,
    pollSnapshotStatus,
    restoreSnapshotTask,
    stopSnapshotPolling,
    getAgentSetting,
    getAgentInfo,
    init,
    snapshotStatus,
    snapshotError,
    snapshotPolling,
  }
})

File: internal/model/setting/setting.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025-2026 lin-snow

package model

import (
	uuidUtil "github.com/lin-snow/ech0/internal/util/uuid"
	"gorm.io/gorm"
)

const (
	EIGHT_HOUR_EXPIRY string = "8_hours"
	ONE_MONTH_EXPIRY  string = "1_month"
	NEVER_EXPIRY      string = "never"
)

// SystemSetting 定义系统设置实体
type SystemSetting struct {
	SiteTitle     string `json:"site_title"`     // 站点标题
	ServerLogo    string `json:"server_logo"`    // 服务器Logo
	ServerName    string `json:"server_name"`    // 服务器名称
	ServerURL     string `json:"server_url"`     // 服务器地址
	AllowRegister bool   `json:"allow_register"` // 是否允许注册'
	DefaultLocale string `json:"default_locale"` // 站点默认语言（如 zh-CN / en-US）
	ICPNumber     string `json:"ICP_number"`     // 备案号
	FooterContent string `json:"footer_content"` // 自定义页脚内容
	FooterLink    string `json:"footer_link"`    // 自定义页脚链接
	MetingAPI     string `json:"meting_api"`     // Meting API 地址
	CustomCSS     string `json:"custom_css"`     // 自定义 CSS
	CustomJS      string `json:"custom_js"`      // 自定义 JS
}

// S3Setting 定义 S3 存储设置实体
type S3Setting struct {
	Enable     bool   `json:"enable"`      // 是否启用 S3 存储
	Provider   string `json:"provider"`    // S3 服务提供商，例如 "aws", "r2", "minio", "other"
	Endpoint   string `json:"endpoint"`    // S3 端点
	AccessKey  string `json:"access_key"`  // 访问密钥 ID
	SecretKey  string `json:"secret_key"`  // 秘密访问密钥
	BucketName string `json:"bucket_name"` // 存储桶名称
	Region     string `json:"region"`      // 区域
	UseSSL     bool   `json:"use_ssl"`     // 是否使用 SSL
	CDNURL     string `json:"cdn_url"`     // CDN 加速域名（可选，没有就走 Endpoint）
	PathPrefix string `json:"path_prefix"` // 存储路径前缀，例如 "uploads/"，方便隔离目录
	PublicRead bool   `json:"public_read"` // 上传时是否默认设置对象为 public-read
	// UsePathStyle 强制使用 path-style 寻址（endpoint/bucket/key）。仅对 provider="other"
	// 生效（aws/minio/r2 由 virefs 预设决定），保存时非 other 会被归零。
	UsePathStyle bool `json:"use_path_style"`
}

// OAuth2Setting 定义 OAuth2 配置结构体
type OAuth2Setting struct {
	Enable       bool     `json:"enable"`        // 是否启用 OAuth2 登录
	Provider     string   `json:"provider"`      // OAuth2 提供商
	ClientID     string   `json:"client_id"`     // OAuth2 Client ID
	ClientSecret string   `json:"client_secret"` // OAuth2 Client Secret
	RedirectURI  string   `json:"redirect_uri"`  // OAuth2 重定向 URI
	Scopes       []string `json:"scopes"`        // OAuth2 请求的权限范围
	AuthURL      string   `json:"auth_url"`      // OAuth2 授权 URL
	TokenURL     string   `json:"token_url"`     // OAuth2 令牌 URL
	UserInfoURL  string   `json:"user_info_url"` // OAuth2 用户信息 URL

	// OIDC 扩展
	IsOIDC  bool   `json:"is_oidc"`  // 是否启用 OIDC
	Issuer  string `json:"issuer"`   // OIDC 颁发者
	JWKSURL string `json:"jwks_url"` // OIDC JWKS URL

	// 认证边界配置（Panel 主配置，ENV 仅默认值）
	AuthRedirectAllowedReturnURLs []string `json:"auth_redirect_allowed_return_urls"`
	CORSAllowedOrigins            []string `json:"cors_allowed_origins"`
}

// PasskeySetting 定义 Passkey(WebAuthn) 配置结构体
type PasskeySetting struct {
	WebAuthnRPID           string   `json:"webauthn_rp_id"`
	WebAuthnAllowedOrigins []string `json:"webauthn_allowed_origins"`
}

// AccessTokenSetting 定义访问令牌设置实体
type AccessTokenSetting struct {
	ID         string `gorm:"type:char(36);primaryKey" json:"id"`         // 访问令牌 ID
	UserID     string `gorm:"type:char(36);index" json:"user_id"`         // 创建该访问令牌的用户 ID
	Token      string `gorm:"type:varchar(255);uniqueIndex" json:"token"` // 访问令牌
	Name       string `json:"name"`                                       // 访问令牌名称
	TokenType  string `gorm:"size:32;index" json:"token_type"`            // 访问令牌类型（access）
	Scopes     string `gorm:"type:text" json:"scopes"`                    // scopes 的 JSON 字符串
	Audience   string `gorm:"size:64;index" json:"audience"`              // token audience
	JTI        string `gorm:"size:64;uniqueIndex" json:"jti"`             // JWT ID
	Expiry     *int64 `json:"expiry"`                                     // 指针类型，NULL 表示永不过期
	LastUsedAt *int64 `json:"last_used_at,omitempty"`                     // 最后一次使用时间
	CreatedAt  int64  `gorm:"autoCreateTime" json:"created_at"`           // 访问令牌创建时间，Unix 秒级时间戳
}

func (a *AccessTokenSetting) BeforeCreate(_ *gorm.DB) error {
	if a.ID == "" {
		a.ID = uuidUtil.MustNewV7()
	}
	return nil
}

// AgentSetting 定义 LLM Agent 设置实体
type AgentSetting struct {
	Enable     bool   `json:"enable"`     // 是否启用 Agent 功能
	Protocol   string `json:"protocol"`   // LLM 接口协议（OpenAI 兼容/Anthropic，OpenAI 兼容覆盖 DeepSeek、Qwen、Ollama 等）
	Model      string `json:"model"`      // LLM 模型名称
	ApiKey     string `json:"api_key"`    // LLM API Key
	Prompt     string `json:"prompt"`     // Agent 额外使用的提示词
	BaseURL    string `json:"base_url"`   // 自定义 API URL（可选）
	Multimodal bool   `json:"multimodal"` // 多模态支持：Chat 检索命中带图 Echo 时，把配图一并递给模型（需所配模型支持视觉）
	// ContextWindow 是模型上下文窗口的 token 数（0=未配置，按保守默认处理）。
	// 用于区间聚合（年终/月度总结）时的取数预算：窗口越大越倾向「一次塞满全部 Echo」，
	// 越小越早转入按月 map-reduce 分层总结。前端以 256k/1m 形式填写、解析成 token 数后存此。
	ContextWindow int `json:"context_window"`
}

type SnapshotSchedule struct {
	Enable         bool   `json:"enable"`          // 是否启用定时快照
	CronExpression string `json:"cron_expression"` // 定时快照的 Cron 表达式
}
